### PR TITLE
Store refactor

### PR DESCRIPTION
**Pull request by lbwexler**
This PR refactors the store to more efficiently reuse its existing records when reloading (or filtering to) identical records.  It also removes the substantial complexity associated with trying to maintain extraneous children and parent pointers in the record state. 

The main target is of course the effective use of `deltaRowMode = true` in ag-Grid.
Still need to do some performance testing to see how much we are picking up, but the results look promising, and this code seems much easier to reason about in any case.

This should also help avoid some workarounds john haynes needed to deal with the old version cloning records between the filtered/unfiltered states.

Decided *not* to tackle nesting the field data within the record on this pass.  We could do that (I had it mostly implemented) but it actually raised some interesting questions.  In particular, it seemed inconvenient that the "record" you get from our context and selection handlers would not be the same "record" you get when implementing ag-Grid tooltips and renderers, etc.

### CODE REVIEW

File: svc/GridExportService.js [hunk 1]
```diff
@@ -127,17 +128,20 @@ export class GridExportService {
         return {data: headers, depth: 0};
     }
 
-    getRecordRowsRecursive(gridModel, records, columns, depth) {
+    getRecordRowsRecursive(gridModel, recordNodes, columns, depth) {
         const {sortBy, treeMode} = gridModel,
-            sortColIds = sortBy.map(it => it.colId),
+            sortFns = sortBy.map(sort => {
```
Comment (haynesjm42) [inline svc/GridExportService.js]: Could be a fairly concise one-liner: `sortFns = sortBy.map(sort => ({colId}) => ({record}) => record[colId])` though might not be as readable

File: data/impl/RecordSet.js [hunk 2]
```diff
@@ -54,63 +51,124 @@ export class RecordSet {
...
-    gatherAllRecords(records, list = [], map = new Map()) {
-        records.forEach(r => {
-            list.push(r);
-            map.set(r.id, r);
-            if (!isEmpty(r.children)) {
-                this.gatherAllRecords(r.children, list, map);
-            }
-        });
-        return {list, map};
+    getParent(rec) {
```
Comment (haynesjm42) [inline data/impl/RecordSet.js]: Any reason not to just have a `parent` prop on Record now that it know about its store?

File: desktop/cmp/store/StoreCountLabel.js [hunk 3]
```diff
@@ -57,7 +58,7 @@ export class StoreCountLabel extends Component {
         if (!store) return null;
 
         const includeChildren = withDefault(this.props.includeChildren, false),
-            count = includeChildren ? store.records.length : store.rootRecords.length,
+            count = includeChildren ? store.count : this.rootCount(store),
```
[the remark this answers is not in the document]
Comment (lbwexler) [reply desktop/cmp/store/StoreCountLabel.js]: Yeah, thought about that, but then we need 'rootCount' and 'allRootCount', and was wondering who would ever want that other than this component.  but maybe this component is enough.  I will do it....thanks

File: data/LocalStore.js [hunk 4]
```diff
@@ -109,49 +135,41 @@ export class LocalStore extends BaseStore {
...
-        rec.children = raw.children ? this.createRecords(raw.children, rec) : [];
-        return rec;
+    /**
+     * Get a flat set of records representing a store in a tree representations
+     *
+     * @param {Record[]} - records
+     *
+     * @returns {[]} -- array of records of form {record, children};
+     */
+    static getRecordsAsTree(records) {
```
[the remark this answers is not in the document]
Comment (lbwexler) [reply data/LocalStore.js]: Same problem as earlier - then we need a getRecordsAsTree() and getAllRecordsAsTree().
But, yea, again, that seems simpler than this approach....

File: data/impl/RecordSet.js (hunk 2 above)
Comment (lbwexler) [reply data/impl/RecordSet.js]: yes, can add a (dynamic) getter for that.  Good idea

File: desktop/cmp/store/StoreCountLabel.js [hunk 5]
```diff
@@ -67,15 +68,21 @@ export class StoreCountLabel extends Component {
...
     getActiveStore() {
         const {gridModel, store} = this.props;
         return store || (gridModel && gridModel.store);
     }
-}
 
+    rootCount(store) {
+        return reduce(
+            store.records,
+            (ret, val) => {if (val.parentId == null) ret++;},
```
[the remark this answers is not in the document]
Comment (lbwexler) [reply desktop/cmp/store/StoreCountLabel.js]: you see how often I use reduce?  Thanks for catching...

**Comment by lbwexler** [comment]
Just to reply to the concern about loadData vs. updateData.  The former will actually replace the existing data (e.g. refresh, full load).  The latter will preserve existing data and just append missing or update existing records...

File: data/LocalStore.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (lbwexler) [reply data/LocalStore.js]: Did this and added new interfaces to BaseStore itself.

**Comment by amcclain** [comment]
Need to look closely at what changed with grid exports

**Comment by lbwexler** [comment]
I can merge this now -- Tom fixed an issue with sorting (that I found as part of the project) that now needs a bit of a merge....

**Pull request merged**